### PR TITLE
feat: add stat function

### PR DESCRIPTION
**Pull request by katcipis**
The function encapsulates the heavy lifting done by benchstat, tried to use as little as possible from it, using default configurations + focusing on the simple scenario of comparing 2 benchmark run sets (benchstat is way more general/flexible).

Trying to capture a simple version of this:

```
benchstat old.txt new.txt
name        old time/op    new time/op    delta
GobEncode     13.6ms ± 1%    11.8ms ± 1%  -13.31%  (p=0.016 n=4+5)
JSONEncode    32.1ms ± 1%    31.8ms ± 1%     ~     (p=0.286 n=4+5)

name        old speed      new speed      delta
GobEncode   56.4MB/s ± 1%  65.1MB/s ± 1%  +15.36%  (p=0.016 n=4+5)
JSONEncode  60.4MB/s ± 1%  61.1MB/s ± 2%     ~     (p=0.286 n=4+5)
```

We can use the delta to compare and fail if it misses the target for a given specific metric (time/space/speed/etc).
After this building block we should be able to finish the tool, we can run benchs, get results, and call this function to provide statistics.

### CODE REVIEW

File: benchcheck_test.go [hunk 1]
```diff
@@ -160,3 +162,92 @@ func TestBenchModuleNoBenchmarks(t *testing.T) {
...
+
+func TestStatBenchmarkResults(t *testing.T) {
+	type testcase struct {
+		name   string
+		oldres []string
+		newres []string
+		want   []benchcheck.StatResult
+	}
+
+	tcases := []testcase{
```
Comment (katcipis) [inline benchcheck_test.go:174]: hopefully we will have more test cases in the future =P

**Comment by i4ki** [comment]
I don't if make sense, but I had to add the command output to the CmdError.Error() in order to understand why my tests were failing:

```
    benchcheck_test.go:173: unexpected error[fail to run: /usr/local/go/bin/go test -bench=. ./... on dir /home/i4k/go/pkg/mod/github.com/!neoway!labs/gis@v0.0.0-20170702081752-562f220cd7ce: exit status 1 (output --- FAIL: TestEncodeFromCSV (0.00s)
            geohash_test.go:87: open /src/github.com/NeowayLabs/gis/encoding/geohash/_testdata/geohash.csv: no such file or directory
        FAIL
        exit status 1
```

```
diff --git a/benchcheck.go b/benchcheck.go
index 2b3bc00..279f10f 100644
--- a/benchcheck.go
+++ b/benchcheck.go
@@ -48,10 +48,11 @@ type CmdError struct {
 // Error returns the string representation of the error.
 func (c *CmdError) Error() string {
        return fmt.Sprintf(
-               "fail to run: %v on dir %s: %v",
+               "fail to run: %v on dir %s: %v (output %s)",
                c.Cmd,
                c.Cmd.Dir,
                c.Err,
+               c.Output,
        )
 }
```

**Comment by katcipis** [comment]
> I don't know if makes sense, but I had to add the command output to the CmdError.Error() in order to understand why my tests were failing:
> 
> ```
>     benchcheck_test.go:173: unexpected error[fail to run: /usr/local/go/bin/go test -bench=. ./... on dir /home/i4k/go/pkg/mod/github.com/!neoway!labs/gis@v0.0.0-20170702081752-562f220cd7ce: exit status 1 (output --- FAIL: TestEncodeFromCSV (0.00s)
>             geohash_test.go:87: open /src/github.com/NeowayLabs/gis/encoding/geohash/_testdata/geohash.csv: no such file or directory
>         FAIL
>         exit status 1
> ```
> 
> ```
> diff --git a/benchcheck.go b/benchcheck.go
> index 2b3bc00..279f10f 100644
> --- a/benchcheck.go
> +++ b/benchcheck.go
> @@ -48,10 +48,11 @@ type CmdError struct {
>  // Error returns the string representation of the error.
>  func (c *CmdError) Error() string {
>         return fmt.Sprintf(
> -               "fail to run: %v on dir %s: %v",
> +               "fail to run: %v on dir %s: %v (output %s)",
>                 c.Cmd,
>                 c.Cmd.Dir,
>                 c.Err,
> +               c.Output,
>         )
>  }
> ```

I usually don't like adding because the output can be quite big and doesn't compose well with other errors in a call chain. What we could do here is to use errors.As on the tests to print the debug information when it fails, including the output (that was my original idea and why we have a CmdError, but I was not doing that on the tests, sorry for that).

**Comment by katcipis** [comment]
@i4ki for showing the cmd output of errors n tests, what about something like this ?

https://github.com/madlambda/benchcheck/pull/8

Example of how the error looks now:

https://github.com/madlambda/benchcheck/runs/6247375851?check_suite_focus=true#step:5:9

**Comment by i4ki** [comment]
cool! I like this way as well.

**Comment by i4ki** [comment]
old:
```
goos: linux
goarch: amd64
cpu: Intel(R) Core(TM) i7-10750H CPU @ 2.60GHz
BenchmarkEncode-12         	11406358	       100.2 ns/op
BenchmarkStripe-12         	1000000000	         0.2726 ns/op
BenchmarkInterleaves-12    	303347204	         3.969 ns/op
PASS
```

new:
```
goos: linux
goarch: amd64
cpu: Intel(R) Core(TM) i7-10750H CPU @ 2.60GHz
BenchmarkEncode-12              	18098190	        61.79 ns/op
BenchmarkEncodeFixedPoint-12    	30849720	        35.11 ns/op
BenchmarkDecode-12              	70118866	        16.54 ns/op
BenchmarkStripe-12              	1000000000	         0.2554 ns/op
BenchmarkInterleaves-12         	303250658	         3.949 ns/op
PASS
ok  	_/home/i4k/go/pkg/mod/github.com/_neoway_labs/gis@v0.0.0-20171205183537-0b758e536cc5/encoding/geohash	5.380s
```
results from my old geohash library (first commit versus optimized in the master).

just sharing maybe to use as testcase.

File: benchcheck.go [hunk 2]
```diff
@@ -3,15 +3,41 @@ package benchcheck
...
+	// List of results for a specific metric
+	BenchResults []BenchResult
+}
+
+// BenchResult is the result showing performance differences
+// for a single benchmark function.
+type BenchResult struct {
+	// Name of the benchmark function
+	Name string
+	// Performance summary on the old benchmark
```
Comment (i4ki) [inline benchcheck.go:33]:
```suggestion
	// Old is the performance summary of the old benchmark.
```

File: benchcheck.go [hunk 3]
```diff
@@ -3,15 +3,41 @@ package benchcheck
...
+	Old string
+	// Performance summary on the new benchmark
```
Comment (i4ki) [inline benchcheck.go:35]:
```suggestion
	// New is the performance summary of the new benchmark.
```

File: benchcheck.go [hunk 4]
```diff
@@ -3,15 +3,41 @@ package benchcheck
...
+	New string
+	// Delta between the old and new performance
```
Comment (i4ki) [inline benchcheck.go:37]:
```suggestion
	// Delta between the old and new performance summaries.
```

File: benchcheck.go [hunk 5]
```diff
@@ -99,3 +133,65 @@ func RunBench(mod Module) ([]string, error) {
...
+	}
+
+	return res
+}
+
+func newBenchResults(rows []*benchstat.Row) []BenchResult {
+	res := make([]BenchResult, len(rows))
+
+	for i, row := range rows {
+		if len(row.Metrics) != 2 {
```
Comment (i4ki) [inline benchcheck.go:178]: is this true even in the case that `new` has benchmarks not present in the `old`?
Eg.: the newer commit introduced a new Bench* test.

Comment (i4ki) [reply benchcheck.go:178]: asking because this is quite common.

**Review by i4ki — APPROVED** [review]
lgtm

File: benchcheck.go (hunk 5 above)
Comment (katcipis) [reply benchcheck.go:178]: will test it

Comment (katcipis) [reply benchcheck.go:178]: Addes some tests validating that this is true (in the sense that it doesn't panic) when there are a few functions that are not present on both, when one of them is empty or when they have benchmarkes but they don't match. all result in empty stats.

**Pull request merged**